pulp_smash/tests/rpm/api_v2/test_rsync_distributor.py

```diff
@@ -0,0 +1,228 @@
+rsync+ssh. The RPM rsync distributor can be used by itself, but the Pulp user
+story and configuration parameters make it clear that the primary use case is
+as a post-distributor. In other words, it's possible to attach both a yum
+distributor and rsync distributor to a repository, and to use only the latter
+for publishes, but most of the interesting use cases arise when both are used
+for publishes.
+
+For more information on the RPM rsync distributor, see `Pulp #1759`_.
+
+.. _Pulp #1759: https://pulp.plan.io/issues/1759
+"""
+from __future__ import unicode_literals
+
+from pulp_smash import api, cli, utils
+from pulp_smash.compat import urljoin, urlparse
+from pulp_smash.constants import REPOSITORY_PATH, RPM_FEED_URL
+from pulp_smash.tests.rpm.api_v2.utils import gen_distributor, gen_repo
+from pulp_smash.tests.rpm.utils import set_up_module as setUpModule  # noqa pylint:disable=unused-import
+
+
+def _make_user(cfg):
+    """A generator to create a user account on the target system.
+
+    Yield a username and private key. The corresponding public key is make the
+    one and only key in the user's ``authorized_keys`` file.
+
+    The username and private key are yielded one-by-one rather than as a pair
+    because the user creation and key creation steps are executed serially.
+    Should the latter fail, the calling function will still be able to delete
+    the created user.
+
+    The user is given a home directory. When deleting this user, make sure to
+    pass ``--remove`` to ``userdel``. Otherwise, the home directory will be
+    left in place.
+    """
+    client = cli.Client(cfg)
+    sudo = '' if utils.is_root(cfg) else 'sudo '
+
+    username = utils.uuid4()[:32]  # See useradd(8)
+    cmd = 'useradd --create-home {0}'
+    client.run((sudo + cmd.format(username)).split())
+    yield username
+
+    cmd = 'runuser -u {0} -- ssh-keygen -N . -f /home/{0}/.ssh/mykey'
+    cmd = (sudo + cmd.format(username)).split()
+    cmd[cmd.index('-N') + 1] = ''
+    client.run(cmd)
+    cmd = 'cp /home/{0}/.ssh/mykey.pub /home/{0}/.ssh/authorized_keys'
+    client.run((sudo + cmd.format(username)).split())
+    cmd = 'cat /home/{0}/.ssh/mykey.pub'
+    private_key = client.run((sudo + cmd.format(username)).split()).stdout
+    yield private_key
+
+
+class _MakeUserMixin(object):  # pylint:disable=too-few-public-methods
+    """A mixin providing the ability to create a temporary.
+
+    This mixin requires that the ``unittest.TestCase`` class from the standard
+    library is a parent class.
+    """
+
+    def make_user(self, cfg):
+        """Create a user and schedule it for deletion.
+
+        :param pulp_smash.config.ServerConfig cfg: Information about the server
+            being targeted.
+        :returns: A ``(username, private_key)`` tuple.
+        """
+        sudo = '' if utils.is_root(cfg) else 'sudo '
+        creator = _make_user(cfg)
+        username = next(creator)
+        self.addCleanup(
```

Comment:
I'm normally wary of adding in these kinds of checks. But it makes more sense here. If `addCleanup` isn't available, I'd rather know before a user is created.